akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -725,9 +725,41 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
         }
       i += 1
     }
+    cleanUpSubstreams(optionalFailureCause)
     setKeepGoing(false)
   }
 
+  private def cleanUpSubstreams(optionalFailureCause: OptionVal[Throwable]): Unit = {
+    if (_subInletsAndOutlets.nonEmpty) {
+      // automatic completion to not leak SubSinkInlets or SubSourceOutlets on stage completion
+      def completeSubInletsAndOutlets(left: List[AnyRef]): Unit = left match {
+        case sub :: tail =>
+          sub match {
+            case inlet: SubSinkInlet[_] =>
+              val subSink = inlet.sink.asInstanceOf[SubSink[_]]
+              optionalFailureCause match {
+                case OptionVal.Some(cause) => subSink.cancelSubstream(cause)
+                case _                     => subSink.cancelSubstream()
+              }
+            case outlet: SubSourceOutlet[_] =>
+              val subSource = outlet.source.asInstanceOf[SubSource[_]]
+              optionalFailureCause match {
+                case OptionVal.Some(cause) => subSource.failSubstream(cause)
+                case _                     => subSource.completeSubstream()
+              }
+            case wat =>
+              throw new IllegalStateException(
+                s"Stage _subInletsAndOutlets contained unexpected element of type ${wat.getClass.toString}")
+          }
+          completeSubInletsAndOutlets(tail)
+        case Nil => // done
+      }
+
+      completeSubInletsAndOutlets(_subInletsAndOutlets)
```

Comment:
That could simply use `foreach` I guess. At least as long as we do the linear scanning to delete elements below, there doesn't seem to be a need to optimize this one-time method.